story: visitor adopts goal

rairezende opened the pull request.

rairezende committed:
Co-authored-by: Saulo Vallory <saulo@toki.life>

rairezende committed:
added properties to user, enrollment and officiation

rairezende committed:
doc: better commented rules and added an email input that must be filled

rairezende committed:
doc: adopt goal entirely redone

svallory commented on libs/atqr/domain/src/lib/domain-spec/api.disl:
```diff
@@ -40,6 +40,7 @@ context Challenges
       @emailConfirmed: boolean
       @playedChallenges: Enrollment[]
       @officiatedChallenges: Officiation[]
+      @organizedChallenges: Owner[]
```
Shouldn't this be a `Challenge[]`?

svallory commented on libs/atqr/domain/src/lib/domain-spec/api.disl:
```diff
@@ -54,11 +55,13 @@ context Challenges
       @pledge: DISL.Money
       @outcome: PlayerOutcome = null
-      @paymentStatus: PaymentStatus = PaymentStatus.Pending
+      @participationStatus: ParticipationStatus.Accepted
```
We should actually remove this and have participation status only in `Invitee`. Please Challenge.@invitees, and createChallenge params `invitees` and `owner` to use the `Invitee` type

svallory commented on libs/atqr/domain/src/lib/domain-spec/stories/playerAdoptsGoal.disl:
```diff
@@ -0,0 +1,27 @@
+read_also ui.disl
+
+entity Goal
+  @persuer: User
```
typo (pursuer)

svallory replied:
You should actually have a `@challenges` property and have `@pursers` be a dynamic property defined by `retun @challenges.map(c => c.enrollments.map(e => e.player)).flat()`

svallory commented on libs/atqr/domain/src/lib/domain-spec/stories/playerAdoptsGoal.disl:
```diff
@@ -0,0 +1,27 @@
+
+entity Goal
+  @persuer: User
+  @goal: Goal
```
This should be a string

svallory commented on libs/atqr/domain/src/lib/domain-spec/stories/playerAdoptsGoal.disl:
```diff
@@ -0,0 +1,27 @@
+
+root Challenge
+  CreateChallenge?
+    AdoptGoal? :email :goal
```
You cannot have a command inside another

svallory commented on libs/atqr/domain/src/lib/domain-spec/stories/playerAdoptsGoal.disl:
```diff
@@ -0,0 +1,27 @@
+      /**
+      * visitor's email must be validated as a player persuing this goal or not through Goal's @persuer
+      */
+    when @email not @persuer
+      CreateChallenge?
+        @email = their email
+        @goal = this Goal
+
+    when @persuer
+      throw "você já persegue esta meta"
+
```
These should be indented in relation to rules

svallory commented on libs/atqr/domain/src/lib/domain-spec/stories/playerAdoptsGoal.disl:
```diff
@@ -0,0 +1,27 @@
+      /**
+      * visitor's email must be validated as a player persuing this goal or not through Goal's @persuer
+      */
+    when @email not @persuer
```
You could have said `when :email not in @pursuser's email`

svallory commented on libs/atqr/domain/src/lib/domain-spec/stories/playerAdoptsGoal.disl:
```diff
@@ -0,0 +1,27 @@
+        @email = their email
+        @goal = this Goal
+
+    when @persuer
```
You need to be more descriptive / precise

rairezende committed:
Update api.disl

changed @organizedChallenges: Owner to @organizedChallenges:Challenge[]

rairezende committed:
Update api.disl

deleted @participationStatus from Enrollment Value Object and changed invitees and owner params values to Invitee[]

rairezende committed:
Delete pre-commit 2

rairezende committed:
Delete updateChallenge.dto 2.ts

rairezende committed:
Delete index 2.ts

rairezende committed:
Delete .yarn-integrity 2

rairezende committed:
Delete index 2.ts

rairezende committed:
Delete challengeRepo.interface 2.ts

rairezende committed:
Delete index 2.ts

rairezende committed:
Delete index 2.ts

rairezende committed:
Delete date-difference 2.js

rairezende committed:
Delete playerRepo.interface 2.ts

rairezende committed:
Delete index 2.ts

rairezende committed:
Update playerAdoptsGoal.disl

corrected the spelling of "pursuer", changed @goal's type to string, added @challenges to Goal entity, bettered text specificity in general.

rairezende committed:
Merge branch 'saulo/domain-definition-disl' into raiane/adopt-challenge-disl

rairezende committed:
deleted duplicate outaded file

rairezende committed:
updated doc to attend both judge and player demands, generalizing them as "visitors"

svallory commented on libs/atqr/domain/src/lib/domain-spec/stories/visitorAdoptsGoal.disl:
```diff
@@ -0,0 +1,28 @@
+read_also enums.disl
+read_also ui.disl
+
+entity Goal
```
You are still missing `domain ATQR` and `context Challenges`

svallory commented on libs/atqr/domain/src/lib/domain-spec/stories/visitorAdoptsGoal.disl:
```diff
@@ -0,0 +1,28 @@
+
+root Challenge
+  CreateChallenge?
+  AdoptGoal? :email :goal
```
If you are creating a Challenge based on a Goal, a Challenge instance cannot be the one to handle the command since it does not exist yet. The goal does. So Goal will need to be another aggregate root, one capable of creating challenges. Btw, maybe all challenge creation should start on the goal (I have to think about that, what do you think?)

rairezende replied:
It does make sense to me.The Goal will end up being kind of a birth place for of the all challenges.

rairezende replied:
It does make sense to me.The Goal will end up being kind of a birth place for of the all challenges.

svallory commented on libs/atqr/domain/src/lib/domain-spec/stories/visitorAdoptsGoal.disl:
```diff
@@ -0,0 +1,28 @@
+    * visitor's email must be validated as a player pursuing or not this goal through Goal's @pursuer
+    */
+    when :email not in @pursuer's email
+      CreateChallenge?
```
This would be a method call. Commands represent user requests to the system. Also, `createChallenge` takes a lot more arguments

rairezende replied:
I changed it to a method call, but kept the two properties below it cause they are the only ones I need to be already guarding values. I imagine the rest of them are already being handled by the createChallenge method.

svallory replied:
There's no "imagine" allowed in coding :) `createChallenge` is a method and can only handle what you pass to it

svallory reviewed (changes requested):
Asked for a few changes, but nothing serious. Great work!

rairezende committed:
changed root aggregate to Goal, added "domain ATQR" and "context Challenges" to the beggining of the document and corrected the second appearence of "CreateChallenge?" to a method call right syntax

svallory commented on libs/atqr/domain/src/lib/domain-spec/stories/visitorAdoptsGoal.disl:
```diff
@@ -0,0 +1,32 @@
+
+context Challenges
+
+  root Goal
```
This needs to go inside an aggregate root which can also be named Goals

svallory commented on libs/atqr/domain/src/lib/domain-spec/stories/visitorAdoptsGoal.disl:
```diff
@@ -0,0 +1,32 @@
+    @challenges: Challenge[]
+    @pursuers: returns @challenges.map(c => c.enrollments.map(e => e.player)).flat()
+
+    CreateChallenge?
```
I don't understand why there's this empty command here

rairezende replied:
I think the idea was lost btwn changes, can't remember why I left this command behind.

svallory commented on libs/atqr/domain/src/lib/domain-spec/stories/visitorAdoptsGoal.disl:
```diff
@@ -0,0 +1,32 @@
+    CreateChallenge?
+    AdoptGoal? :email :goal
+
+    rules:
+      /**
+      * visitor's email must be validated as a player pursuing or not this goal through Goal's @pursuer
+      */
+      when :email not in @pursuer's email
+        createChallenge
+          @email = visitor's :email input
+          @goal = visitor's :goal of choice
+
+      when @pursuer
+        throw "o usuário já persegue esta meta"
+
+    end rules
```
Shouldn't these be inside `AdoptGoal?`. Also, this command needs to carry all the information necessary to create a `Challenge`. You could probably add a command `isPursuer?` that simply validate the email is not among pursuers

rairezende replied:
I changed AdoptGoal?'s name to IsPursuer? cause it makes more sense that way, since this command is about checking exactly if the email belongs to a already pursuer of the goal. But for me IsPursuer? only needs the params that help with this checking and the rest is a job for "createChallenge" method call which is called after in this file.

svallory replied:
You are correct. And it MUST NOT create a challenge in the process of checking if the pursuer is there. That's why I said "add a isPurser command" and not rename :) You need both

rairezende committed:
deleted uneened command "CreateChallenge?" and fixed indentation

rairezende committed:
changed AdoptGoal? for IsPersuer? since its meaning makes more sense.

rairezende committed:
fixed "pursuer" ortography

rairezende committed:
put back createChallenge? command instead of createChallenge method call

rairezende committed:
deleted the method inside isPursuer and put it after the end of rules

rairezende committed:
added params to AdoptChallenge that now is called AdoptGoal

rairezende committed:
deleted uneeded space

svallory committed:
Final review and fixes to api.disl
